File: src/common-py/secrets_client_py/secrets_client/_tls.py
"""
Hot-swapping TLS context for the Python secrets-client library.

Reads three PEM files — cert.pem, chain.pem, key.pem — from a directory and
builds an ssl.SSLContext. A rotation event on any of the three files schedules
a coalesced reload (within coalesce_ms) so that three rapid renames produce
exactly one context swap.

File semantics:
  cert.pem  — leaf server certificate (required).
  key.pem   — private key matching cert.pem (required).
  chain.pem — intermediate CA certificates to append to cert.pem so peers can
              verify the full chain up to a trusted root. Optional; omit or
              leave empty when the leaf cert is directly signed by a root CA or
              when intermediates are already concatenated into cert.pem.

The swap is a single reference assignment under a threading.Lock, ensuring
callers of `context()` always see a consistent, fully-assembled SSLContext.
"""

from __future__ import annotations

import os
import ssl
import tempfile
import threading
from pathlib import Path
from typing import Callable


def _build_ssl_context(cert_dir: Path) -> ssl.SSLContext:
    cert_path = cert_dir / "cert.pem"
    chain_path = cert_dir / "chain.pem"
    key_path = cert_dir / "key.pem"

    ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)

    chain_bytes = b""
    if chain_path.exists() and chain_path.stat().st_size > 0:
        chain_bytes = chain_path.read_bytes()

    if chain_bytes:
        # ssl.SSLContext.load_cert_chain requires a file path, not bytes, so
        # concatenate the leaf cert + intermediates into a temporary file.
        combined = cert_path.read_bytes() + b"\n" + chain_bytes
        tmp_fd, tmp_path = tempfile.mkstemp(suffix=".pem")
        try:
            with os.fdopen(tmp_fd, "wb") as tmp_file:
                tmp_file.write(combined)
            ctx.load_cert_chain(certfile=tmp_path, keyfile=str(key_path))
        finally:
            os.unlink(tmp_path)
    else:
        ctx.load_cert_chain(certfile=str(cert_path), keyfile=str(key_path))

    return ctx


class ReloadableSSLContext:
    """
    Wraps an ssl.SSLContext that is atomically hot-swapped on TLS rotation.

    Parity with the TypeScript `TlsContext` interface.
    """

    def __init__(
        self,
        cert_dir: Path,
        coalesce_ms: int,
        on_error: Callable[[str, Exception], None],
    ) -> None:
        self._cert_dir = cert_dir
        self._coalesce_ms = coalesce_ms
        self._on_error = on_error
        self._lock = threading.Lock()
        self._current: ssl.SSLContext = _build_ssl_context(cert_dir)
        self._subscribers: list[Callable[[], None]] = []
        self._timer: threading.Timer | None = None

    def context(self) -> ssl.SSLContext:
        """Returns the current SSLContext; always reflects the latest rotation."""
        with self._lock:
            return self._current

    def on_rotate(self, cb: Callable[[], None]) -> Callable[[], None]:
        """Subscribe to rotation. Returns an unsubscribe callable."""
        with self._lock:
            self._subscribers.append(cb)

        def _unsub() -> None:
            with self._lock:
                try:
                    self._subscribers.remove(cb)
                except ValueError:
                    pass

        return _unsub

    def schedule_reload(self) -> None:
        """Signal that one of the three TLS files may have changed."""
        with self._lock:
            if self._timer is not None:
                self._timer.cancel()
            delay = self._coalesce_ms / 1000.0
            self._timer = threading.Timer(delay, self._reload)
            self._timer.daemon = True
            self._timer.start()

    def _reload(self) -> None:
        try:
            new_ctx = _build_ssl_context(self._cert_dir)
            with self._lock:
                self._current = new_ctx
                subs = list(self._subscribers)
            for cb in subs:
                try:
                    cb()
                except Exception as err:
                    self._on_error(f"tls:{self._cert_dir.name}:on_rotate", err)
        except Exception as err:
            self._on_error(f"tls:{self._cert_dir.name}", err)

    def close(self) -> None:
        with self._lock:
            if self._timer is not None:
                self._timer.cancel()
                self._timer = None
            self._subscribers.clear()

File: src/common-rs/secrets-client/src/tls.rs
//! Hot-swapping TLS context for the Rust secrets-client library.
//!
//! Reads three PEM files — cert.pem, chain.pem, key.pem — from a directory
//! and builds a `rustls::ServerConfig`. A rotation event on any of the three
//! files schedules a coalesced reload (within `coalesce_ms`) so that three
//! rapid renames produce exactly one context swap.
//!
//! The swap is a single `arc_swap::ArcSwap` store — readers calling
//! `current()` always see a consistent, fully-assembled `ServerConfig`.

use std::{
    collections::HashMap,
    fs,
    io::BufReader,
    path::PathBuf,
    sync::{mpsc, Arc, Mutex, Weak},
    thread,
    time::Duration,
};

use arc_swap::ArcSwap;
use rustls::{
    pki_types::{CertificateDer, PrivateKeyDer},
    ServerConfig,
};
use rustls_pemfile::{certs, private_key};

use crate::SecretError;

/// Builds a `rustls::ServerConfig` from the three PEM files in `cert_dir`.
///
/// - `cert.pem`  — leaf server certificate (required).
/// - `key.pem`   — private key matching cert.pem (required).
/// - `chain.pem` — intermediate CA certificates appended to cert.pem.
///   Optional; ENOENT is silently ignored.
pub(crate) fn build_server_config(
    cert_dir: &std::path::Path,
) -> Result<Arc<ServerConfig>, SecretError> {
    // Ensure the ring crypto provider is installed. This call is idempotent;
    // if another call already installed it, the Err is silently ignored.
    let _ = rustls::crypto::ring::default_provider().install_default();
    let cert_pem =
        fs::read(cert_dir.join("cert.pem")).map_err(|e| SecretError::io("cert.pem", e))?;
    let key_pem = fs::read(cert_dir.join("key.pem")).map_err(|e| SecretError::io("key.pem", e))?;

    let mut combined_cert_pem = cert_pem;
    let chain_path = cert_dir.join("chain.pem");
    match fs::read(&chain_path) {
        Ok(chain) if !chain.is_empty() => {
            combined_cert_pem.push(b'\n');
            combined_cert_pem.extend_from_slice(&chain);
        }
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {}
        Err(e) => return Err(SecretError::io("chain.pem", e)),
        _ => {}
    }

    let cert_list: Vec<CertificateDer<'static>> = {
        let mut reader = BufReader::new(combined_cert_pem.as_slice());
        certs(&mut reader)
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| SecretError::Other(format!("parsing cert.pem: {e}")))?
    };

    let key: PrivateKeyDer<'static> = {
        let mut reader = BufReader::new(key_pem.as_slice());
        private_key(&mut reader)
            .map_err(|e| SecretError::Other(format!("parsing key.pem: {e}")))?
            .ok_or_else(|| SecretError::Other("no private key found in key.pem".into()))?
    };

    let config = ServerConfig::builder()
        .with_no_client_auth()
        .with_single_cert(cert_list, key)
        .map_err(|e| SecretError::Other(format!("building ServerConfig: {e}")))?;

    Ok(Arc::new(config))
}

/// Hot-swapping TLS context backed by three PEM files in a directory.
pub struct ReloadableTlsContext {
    cert_dir: PathBuf,
    on_error: Arc<crate::OnErrorFn>,
    current: ArcSwap<ServerConfig>,
    // Store subscribers as Arc<dyn Fn()> so we can clone them out of the map
    // before calling, avoiding holding the lock during callback invocation.
    subs: Mutex<HashMap<u64, Arc<dyn Fn() + Send + Sync + 'static>>>,
    next_id: Mutex<u64>,
    // Sending on this channel requests a coalesced reload.
    reload_tx: mpsc::Sender<()>,
}

impl ReloadableTlsContext {
    pub(crate) fn new(
        cert_dir: PathBuf,
        coalesce_ms: u64,
        on_error: Arc<crate::OnErrorFn>,
    ) -> Result<Arc<Self>, SecretError> {
        let config = build_server_config(&cert_dir)?;
        let (reload_tx, reload_rx) = mpsc::channel::<()>();
        let coalesce = Duration::from_millis(coalesce_ms);

        // Use a Weak slot to give the worker a back-reference without creating
        // an Arc cycle. The slot is populated immediately after Arc::new; no
        // schedule_reload() can fire before new() returns.
        let weak_slot: Arc<Mutex<Option<Weak<ReloadableTlsContext>>>> = Arc::new(Mutex::new(None));
        let weak_slot_for_worker = Arc::clone(&weak_slot);

        thread::spawn(move || {
            loop {
                // Block until at least one reload is requested.
                if reload_rx.recv().is_err() {
                    break;
                }
                // Drain additional rapid requests within the coalesce window.
                loop {
                    match reload_rx.recv_timeout(coalesce) {
                        Ok(()) => continue,
                        Err(mpsc::RecvTimeoutError::Timeout) => break,
                        // Channel closed during coalesce — skip reload and exit.
                        Err(mpsc::RecvTimeoutError::Disconnected) => return,
                    }
                }
                if let Some(weak) = weak_slot_for_worker.lock().unwrap().as_ref() {
                    if let Some(ctx) = weak.upgrade() {
                        ctx.reload();
                    }
                }
            }
        });

        let ctx = Arc::new(Self {
            cert_dir,
            on_error,
            current: ArcSwap::from(config),
            subs: Mutex::new(HashMap::new()),
            next_id: Mutex::new(0),
            reload_tx,
        });

        // Populate the slot now that the Arc exists.
        *weak_slot.lock().unwrap() = Some(Arc::downgrade(&ctx));

        Ok(ctx)
    }

    /// Returns the current `ServerConfig`; always reflects the latest rotation.
    pub fn current(&self) -> Arc<ServerConfig> {
        self.current.load_full()
    }

    /// Registers a callback invoked after each successful context swap.
    /// Returns an unsubscribe function.
    pub fn on_rotate(self: &Arc<Self>, cb: impl Fn() + Send + Sync + 'static) -> impl Fn() {
        let cb: Arc<dyn Fn() + Send + Sync + 'static> = Arc::new(cb);
        let mut id_guard = self.next_id.lock().unwrap();
        let id = *id_guard;
        *id_guard += 1;
        drop(id_guard);
        self.subs.lock().unwrap().insert(id, cb);

        let weak = Arc::downgrade(self);
        move || {
            if let Some(ctx) = weak.upgrade() {
                ctx.subs.lock().unwrap().remove(&id);
            }
        }
    }

    /// Schedules a coalesced TLS reload. Rapid calls within `coalesce_ms`
    /// collapse to a single rebuild.
    pub(crate) fn schedule_reload(&self) {
        let _ = self.reload_tx.send(());
    }

    fn reload(&self) {
        match build_server_config(&self.cert_dir) {
            Ok(next) => {
                self.current.store(next);
                // Snapshot Arc-cloned subscribers before calling to avoid
                // holding the lock during callback invocation.
                let cbs: Vec<Arc<dyn Fn() + Send + Sync>> =
                    self.subs.lock().unwrap().values().cloned().collect();

                let name = self
                    .cert_dir
                    .file_name()
                    .unwrap_or_default()
                    .to_string_lossy()
                    .into_owned();
                for cb in cbs {
                    if let Err(e) = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| cb()))
                    {
                        (self.on_error)(
                            &format!("tls:{name}:on_rotate"),
                            &SecretError::Other(format!("subscriber panic: {e:?}"))
                                as &dyn std::error::Error,
                        );
                    }
                }
            }
            Err(e) => {
                let name = self
                    .cert_dir
                    .file_name()
                    .unwrap_or_default()
                    .to_string_lossy();
                (self.on_error)(&format!("tls:{name}"), &e as &dyn std::error::Error);
                // current retains last-known-good
            }
        }
    }
}
